scripts/data/src/org/akvo/gae/remoteapi/FixOldTranslations.java

```diff
@@ -0,0 +1,129 @@
+            case "QUESTION_TEXT":
+            case "QUESTION_NAME":
+            case "QUESTION_DESC":
+            case "QUESTION_TIP":
+                fixed = fixQuestionTranslation(t, ds);
+                break;
+            case "QUESTION_GROUP_NAME":
+            case "QUESTION_GROUP_DESC":
+                fixed = fixQuestionGroupTranslation(t, ds);
+                break;
+            case "QUESTION_OPTION":
+                fixed = fixQuestionOptionTranslation(t, ds);
+                break;
+            default:
+                break;
+        }
+        return fixed;
+    }
+
+    private Entity fixQuestionGroupTranslation(Entity t, final DatastoreService ds) {
+        System.out.println("Fixing QuestionGroup translation");
+        Long parentId = (Long) t.getProperty("parentId");
+        return fixTranslation(t, ds, parentId);
+    }
+
+    private Entity fixQuestionTranslation(Entity t, final DatastoreService ds) {
+        System.out.println("Fixing Question translation");
+        Long parentId = (Long) t.getProperty("parentId");
+        if (parentId != null) {
+            final Entity question = fetchQuestion(ds, parentId);
+            if (question != null) {
+                return fixTranslation(t, ds, (Long) question.getProperty("questionGroupId"));
+            }
+        }
+        return null;
+    }
+
+    private Entity fixQuestionOptionTranslation(Entity t, final DatastoreService ds) {
+        System.out.println("Fixing QuestionOption translation");
+        Long parentId = (Long) t.getProperty("parentId");
+        if (parentId != null) {
+            final Query.Filter fs = new Query.FilterPredicate(Entity.KEY_RESERVED_PROPERTY,
+                    Query.FilterOperator.EQUAL, KeyFactory.createKey("QuestionOption", parentId));
+            final Query q = new Query("QuestionOption").setFilter(fs);
+            Entity option = ds.prepare(q).asSingleEntity();
+            if (option != null) {
+                final Entity question = fetchQuestion(ds, (Long) option.getProperty("questionId"));
+                if (question != null) {
+                    return fixTranslation(t, ds, (Long) question.getProperty("questionGroupId"));
+                }
+            }
+        }
+        return null;
+    }
+
+    private Entity fixTranslation(Entity t, DatastoreService ds, Long groupId) {
+        final Entity qGroup = fetchGroup(ds, groupId);
+        if (qGroup != null) {
+            t.setProperty("surveyId", qGroup.getProperty("surveyId"));
+            t.setProperty("questionGroupId", groupId);
+            return t;
+        }
+        return null;
+    }
+
+    private List<Entity> fetchBrokenTranslations(DatastoreService ds) {
+        List<Entity> translations = new ArrayList<>();
+        final Query q = new Query("Translation");
+        final PreparedQuery pq = ds.prepare(q);
+        final Query.Filter filterNullSurveyId = new Query.FilterPredicate("surveyId", Query.FilterOperator.EQUAL, null);
```

Comment:
I wonder if this query is *"correct"*. In the Datastore a *missing* property gets exposed as a `null` value, when you have already the object in memory, but you can't query for entities with a missing property. See: https://cloud.google.com/appengine/docs/standard/java/datastore/query-restrictions#entities_lacking_a_property_named_in_the_query_are_ignored

If we're sure that the value **is** `null` then just ignore this comment.

Reply:
yes I am sure this works, I have tested it and Translations do get found with these missing properties and loads of them